### ISSUE TITLE
Different results using generator with multi-nodes/GPUs or with single GPU

### ISSUE THREAD

**Issue by shinechou** [uber/horovod#89]
@alsrgv: thank you so much for providing such a nice example. Now I can run my script with multi-nodes/GPUs, but I have a few questions about it because I have noticed the difference of the results when I run my test with multi-nodes/GPUs when I use generator. I didn't notice the difference if I don't use generator. 

I have noticed that in your example, you divide the number of batches by hvd.size(), it works fine in my case. But in your keras_mnist_advanced example when you use generator, you divide the steps_per_epoch by hvd.size(), then I noticed different results with my own test data using generator, the training and validation accuracy look like zigzag for all the epochs, but with single GPU then it is not like zigzag. What could be the reason? Or what could I try further to get rid of this difference?

Another question, about the "validation_steps=3 * test_batches // hvd.size()", how did you decide this factor of 3? I tried both 3 and 1, the results are also different, but both show the zigzag accuracy results with multi-nodes/GPUs. And a very strange observation is, the validation accuracy is very high in the first few epochs and I'm monitoring the val_acc to decide whether I'll save the trained weights for later use, in this case I always save the trained weights at the first or second epoch because the validation accuracy is always lower than that after 70/80 epochs.

Thanks a lot in advance.

**Comment by alsrgv** [comment]
@shinechou, what is the size of your test data set?  Factor `3` was empirical.  Can you try removing division for the `validation_steps`, and just set `validation_steps = test_batches`?

I do use this same scheme for ImageNet training experiments and it seems to be working pretty well.

**Comment by shinechou** [comment]
@alsrgv: thanks. Yes, as I mentioned, I tried "validation_steps = test_batches" and it also shows the zigzag accuracy results. But if I don't use generator but just load all the data (training/validation/test data sets) into RAM, the results are going well then. For my data, it is not too big, 256 classes, 960 training samples  and 240 validation samples per class, each sample is a 744x1(width x height) pixel curve. 30000 curves of test data. In total it is less than 1G of data, but I need to handle much bigger data in my real experiments that's why I wanna use generator.

**Comment by alsrgv** [comment]
@shinechou, another thing that is different about generator example is the presence of the warmup callback. I recommend upgrading to 0.11.0 which was just released to take the advantage of it.

You typically want to scale your learning rate by the number of GPUs - `LR * had.size()`. However, if you scale it immediately, your final accuracy may suffer. Facebook purposed to use smaller learning rate in the beginning and linearly scale it for the first few epochs. `LearningRateWarmupCallback` shipped with 0.11.0 does that. See [updated example](https://github.com/uber/horovod/blob/master/examples/keras_mnist_advanced.py) for details.

**Comment by shinechou** [comment]
@alsrgv: thank you very much for your improved version! I'll try it out and let you know the results. I have a question about the initial learning rate "opt = keras.optimizers.Adadelta(lr=1.0 * hvd.size())", when I use another optimizer I assume that I need to change the "1.0" to default learning rate of that optimizer accordingly, right?

**Comment by alsrgv** [comment]
@shinechou, that's right!

**Comment by alsrgv** [comment]
Closing this issue due to lack of activity.  Please feel free to reopen or raise a new one if you have more questions.
